libtransact/src/workload/runner.rs

```diff
@@ -0,0 +1,425 @@
+            debug!("Shutting down worker {}", worker.id);
+            if worker.sender.send(ShutdownMessage).is_err() {
+                return Err(WorkloadRunnerError::WorkloadRemoveError(format!(
+                    "Failed to send shutdown messages to {}",
+                    id,
+                )));
+            }
+
+            if let Some(thread) = worker.thread.take() {
+                if let Err(err) = thread.join() {
+                    return Err(WorkloadRunnerError::WorkloadRemoveError(format!(
+                        "Failed to cleanly join worker thread {}: {:?}",
+                        id, err,
+                    )));
+                }
+            }
+        } else {
+            return Err(WorkloadRunnerError::WorkloadRemoveError(format!(
+                "Workload with ID {} does not exist",
+                id,
+            )));
+        }
+
+        Ok(())
+    }
+
+    /// Shutsdown all running workloads
+    pub fn shutdown(self) {
+        for (_, worker) in self.workloads.iter() {
+            if worker.sender.send(ShutdownMessage).is_err() {
+                warn!("Failed to send shutdown messages to {}", worker.id);
+            }
+        }
+
+        for (_, mut worker) in &mut self.workloads.into_iter() {
+            debug!("Shutting down worker {}", worker.id);
+            if let Some(thread) = worker.thread.take() {
+                if let Err(_err) = thread.join() {
+                    warn!("Failed to cleanly join worker thread {}", worker.id);
+                }
+            }
+        }
+    }
+}
+
+/// Sent to a workload to signal it should stop
+struct ShutdownMessage;
+
+/// Represents a running workload
+struct Worker {
+    id: String,
+    thread: Option<thread::JoinHandle<()>>,
+    sender: Sender<ShutdownMessage>,
+}
+
+#[derive(Default)]
+struct WorkerBuilder {
+    id: Option<String>,
+    workload: Option<Box<dyn BatchWorkload>>,
+    targets: Option<Vec<String>>,
+    rate: Option<u64>,
+    auth: Option<String>,
+}
+
+impl WorkerBuilder {
+    /// Sets the ID of the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `id` - The unique ID of the worker
+    pub fn with_id(mut self, id: &str) -> WorkerBuilder {
+        self.id = Some(id.into());
+        self
+    }
+
+    /// Sets the workload that will be run against the targets
+    ///
+    /// # Arguments
+    ///
+    ///  * `workload` - The workload that will return batches to submit
+    pub fn with_workload(mut self, workload: Box<dyn BatchWorkload>) -> WorkerBuilder {
+        self.workload = Some(workload);
+        self
+    }
+
+    /// Sets the targets for the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `targets` - A list of URL for submitting the batches. The URL provided must be the full
+    ///               URL before adding `/batches` for submission
+    pub fn with_targets(mut self, targets: &[String]) -> WorkerBuilder {
+        self.targets = Some(targets.to_vec());
```

Comment:
The argument should just be `Vec<String>` to prevent an unnecessary clone. If the calling code doesn't need the vec anymore, it can just pass ownership; it should be up to the caller to determine if a clone is needed.

Reply:
This is true with the methods that take `&str` as well - a builder is the equivalent of a constructor, so it should really should indicate that it is taking ownership of the values (which it will ultimately pass to the final constructed struct).

Reply:
updated